src/main/java/de/tum/in/www1/artemis/service/connectors/AbstractContinuousIntegrationService.java

```diff
@@ -69,7 +74,48 @@ public Result createResultFromBuildResult(AbstractBuildResultNotificationDTO bui
      * @param result      the result for which the feedback should be added
      * @param buildResult The build result
      */
-    protected abstract void addFeedbackToResult(Result result, AbstractBuildResultNotificationDTO buildResult);
+    private void addFeedbackToResult(Result result, AbstractBuildResultNotificationDTO buildResult) {
+        final var jobs = buildResult.getBuildJobs();
+        final var programmingExercise = (ProgrammingExercise) result.getParticipation().getExercise();
+        final var programmingLanguage = programmingExercise.getProgrammingLanguage();
+        final var projectType = programmingExercise.getProjectType();
+
+        for (final var job : jobs) {
+            // 1) add feedback for failed test cases
+            for (final var failedTest : job.getFailedTests()) {
+                result.addFeedback(feedbackRepository.createFeedbackFromTestCase(failedTest.getName(), failedTest.getMessage(), false, programmingLanguage, projectType));
+            }
+            result.setTestCaseCount(result.getTestCaseCount() + job.getFailedTests().size());
+
+            // 2) add feedback for passed test cases
+            for (final var successfulTest : job.getSuccessfulTests()) {
+                result.addFeedback(feedbackRepository.createFeedbackFromTestCase(successfulTest.getName(), successfulTest.getMessage(), true, programmingLanguage, projectType));
+            }
+            result.setTestCaseCount(result.getTestCaseCount() + job.getSuccessfulTests().size());
+            result.setPassedTestCaseCount(result.getPassedTestCaseCount() + job.getSuccessfulTests().size());
+        }
+
+        // 3) process static code analysis feedback
+        final var staticCodeAnalysisReports = buildResult.getStaticCodeAnalysisReports();
+        if (Boolean.TRUE.equals(programmingExercise.isStaticCodeAnalysisEnabled()) && staticCodeAnalysisReports != null && !staticCodeAnalysisReports.isEmpty()) {
+            var scaFeedbackList = feedbackRepository.createFeedbackFromStaticCodeAnalysisReports(staticCodeAnalysisReports);
+            result.addFeedbacks(scaFeedbackList);
+            result.setCodeIssueCount(scaFeedbackList.size());
+        }
+
+        // 4) process testwise coverage analysis report
+        if (Boolean.TRUE.equals(programmingExercise.isTestwiseCoverageEnabled())) {
+            var report = buildResult.getTestwiseCoverageReports();
+            if (report != null) {
+                // since the test cases are not saved to the database yet, the test case is null for the entries
+                var coverageFileReportsWithoutTestsByTestCaseName = testwiseCoverageService.createTestwiseCoverageFileReportsWithoutTestsByTestCaseName(report);
+                result.setCoverageFileReportsByTestCaseName(coverageFileReportsWithoutTestsByTestCaseName);
+            }
+        }
```

Comment:
What do you think about moving the three blocks `for (job : jobs)`, 3), 4) each into their own method? They seem to be fairly independent of each other.

```suggestion
        addTestCaseFeedbacksToResult(result, jobs, programmingLanguage, projectType);
        addStaticCodeAnalysisFeedbackToResult(result, buildResult, programmingExercise);
        addTestwiseCoverageReportToResult(result, buildResult, programmingExercise);
```